includes/Classifai/Providers/Azure/Personalizer.php

```diff
@@ -0,0 +1,604 @@
+				$markup .= sprintf(
+					'<div class="wp-block-classifai-recommended-content__post-author">%1$s</div>',
+					esc_html( $byline )
+				);
+			}
+		}
+
+		if ( isset( $attributes['displayPostDate'] ) && $attributes['displayPostDate'] ) {
+			$markup .= sprintf(
+				'<time datetime="%1$s" class="wp-block-classifai-recommended-content__post-date">%2$s</time>',
+				esc_attr( get_the_date( 'c', $post ) ),
+				esc_html( get_the_date( '', $post ) )
+			);
+		}
+
+		if ( isset( $attributes['displayPostExcerpt'] ) && $attributes['displayPostExcerpt'] ) {
+			$trimmed_excerpt = get_the_excerpt( $post );
+
+			if ( post_password_required( $post ) ) {
+				$trimmed_excerpt = __( 'This content is password protected.', 'classifai' );
+			}
+
+			$markup .= sprintf(
+				'<div class="wp-block-classifai-recommended-content__post-excerpt">%1$s</div>',
+				esc_html( $trimmed_excerpt )
+			);
+		}
+
+		$markup .= "</li>\n";
+
+		$class = 'wp-block-classifai-recommended-content';
+
+		if ( isset( $attributes['displayPostDate'] ) && $attributes['displayPostDate'] ) {
+			$class .= ' has-dates';
+		}
+
+		if ( isset( $attributes['displayAuthor'] ) && $attributes['displayAuthor'] ) {
+			$class .= ' has-author';
+		}
+		return sprintf(
+			'<ul class="%1$s">%2$s</ul>',
+			esc_attr( $class ),
+			$markup
+		);
+	}
+
+	/**
+	 * Get Features for specific post.
+	 *
+	 * @param int $post_id Post Object.
+	 * @return array
+	 */
+	protected function get_post_features( $post_id ) {
+		$features = array(
+			'title'   => $this->get_string_words( get_the_title( $post_id ) ),
+			'excerpt' => $this->get_string_words( get_the_excerpt( $post_id ) ),
+			'_URL'    => get_the_permalink( $post_id ),
+		);
+
+		$post_type = get_post_type( $post_id );
+
+		// Get post type taxonomies.
+		$taxonomies = get_object_taxonomies( $post_type, 'objects' );
+		foreach ( $taxonomies as $taxonomy_slug => $taxonomy ) {
+			// Get the terms related to post.
+			$terms = get_the_terms( $post_id, $taxonomy_slug );
+
+			if ( ! empty( $terms ) ) {
+				foreach ( $terms as $term ) {
+					$features[ $taxonomy->label ][ $term->name ] = 1;
+				}
+			}
+		}
+
+		return $features;
+	}
+
+	/**
+	 * Get user agent for personilizer contextFeatures.
+	 */
+	protected function get_user_agent_features() {
+		// phpcs:ignore WordPressVIPMinimum.Variables.RestrictedVariables.cache_constraints___SERVER__HTTP_USER_AGENT__
+		$user_agent = isset( $_SERVER['HTTP_USER_AGENT'] ) ? sanitize_text_field( wp_unslash( $_SERVER['HTTP_USER_AGENT'] ) ) : '';
+
+		// User Agent Parsing
+		$parser    = Parser::create();
+		$ua_result = $parser->parse( $user_agent );
+
+		return array(
+			'_ua'           => $ua_result->originalUserAgent, // phpcs:ignore WordPress.NamingConventions.ValidVariableName.UsedPropertyNotSnakeCase
+			'_DeviceFamily' => $ua_result->device->family,
+			'_OSFamily'     => $ua_result->os->family,
+			'DeviceBrand'   => $ua_result->device->brand,
+			'DeviceModel'   => $ua_result->device->model,
```

Comment:
Do we need any escaping here?

Reply:
I think it is fine to use this without escaping, as these data come from parser lib and we are not storing or showing it anywhere, just sending request to the personalizer with this data.